### ISSUE TITLE
Should Consumer.messages be flushed after rebalance?

### ISSUE THREAD

**Issue by jasonparekh** [bsm/sarama-cluster#22]
(Applies to the v2 branch)

I imagine a problem where a rebalance occurs and the partitions that consumer A was assigned are now given to consumer B, but consumer A's Consumer.messages channel still has some messages queued from before the rebalance.

I was thinking it may be possible to flush in the app business logic given there is a rebalance notification, but by the time that is delivered, the consumer loop logic has started executing and pushing messages from the post-rebalance partitions onto this same channel.

Thoughts?  I understand that this library can't achieve exactly-once delivery semantics on its own, but fixing this could reduce the window of potentially processing a message multiple times.

**Comment by dim** [comment]
So the way I was trying to account for that is via https://github.com/bsm/sarama-cluster/blob/13cdf089dcdaac63234b50af55f49e164f702b95/consumer.go#L208

What do you think?

**Comment by jasonparekh** [comment]
Cool, glad to hear it was considered.

That's going to be too short of a sleep duration for most users since by default the messages channel has a buffer of 256, and that MaxProcessingTime is meant per-message, so it could take up to 25.6s to flush the buffer by letting the app process the messages.

In my case, we have messages that take significantly longer to process (minutes), but have the ability to abort.  So, if I were told about an upcoming rebalance before the new messages were pushed into the messages channel, I could abort all processing and flush the messages myself.

The implementation is almost what I need, the only issue is the notification is sent via a `defer` in `rebalance`, whereas the goroutines to pump into the `messages` channel are launched from within the `rebalance` (via `createConsumer`).

**Comment by tcrayford** [comment]
@jasonparekh @dim I don't think sending the notification immediately in `rebalance` works quite that simply. The issue is that we call `release()` in `rebalance()` before we rejoin the group. I think however, an acceptable workaround would be to send two notifications during rebalancing: once for leaving, then once again for the joining. I'm happy to send a PR for this at some point soon.

However, this makes the client notably more annoying to use correctly - you have to do a `select` over the `Messages()` and the `Notifications()` channel, and track subscription yourself as well (and discard messages once the `Notification` informs you that you no longer have that subscription). I don't have a good remedy for that yet - it seems endemic in the channel based interface `sarama` and `sarama-cluster` embody (in the JVM consumer, all of these issues are avoided by having a synchronous batch based polling approach).

**Comment by dim** [comment]
@jasonparekh @tcrayford the cluster consumer could keep track of 'emitted' offsets too and be aware of messages that have been sent to the `Messages()` channel. On rebalance it then could:

1. stop consumers and stop pushing new messages into the messages channel
2. wait for a configurable period of time until the messages channel is fully drained
3. wait for a configurable period of time until the stored emitted offsets match the processed (marked) offsets
4. release all held partitions, rejoin group, subscribe to new partitions, etc

Notifications should realistically only be used for logging and monitoring.

@jasonparekh I don't think it makes sense to abort on rebalance. Once a rebalance cycle kicks in, the app should be given reasonable time to complete the processing. If the processing of your messages takes minutes, I would nevertheless recommend to adapt `MaxProcessingTime` to reflect your requirements, but also decrease `ChannelBufferSize`, as you also want to prevent large queues.

**Comment by jasonparekh** [comment]
@dim I generally agree that aborting isn't good practice, but we have large variance in processing times (which isn't ideal, but unfortunately don't have time to clean up ATM).

For the project, I agree it's better to go that route than employing another notification.  How quickly do you think that could land?

**Comment by dim** [comment]
I will take a look in the next few days, if I get the time, but I cannot promise you anything

**Comment by tcrayford** [comment]
Hi,

I think I'm a bit more comfortable with this use of sleep these days. However, it causes us an issue. For our use case, when a consumer is first assigned a partition, we have to load quite a bit of state from external storage, which sometimes takes between 30s and a few minutes. However, once we've reloaded state, our messages process very quickly (typically under 1ms/message). So, for state reloading we want to set sarama's Max processing time *very* high - right now we set it to 5 minutes. But, this means on every rebalance we suffer at least 5 minutes of downtime, even if actually in this case state loading took 1 second (which it can if you're reassigned a partition you already loaded state for). Would you be interested in using a separate config var for the sleep times in sarama-cluster? I am happy writing the code/etc. With that, we could use sarama-cluster from here rather than a fork.

We listen to sarama cluster notifications during state reloading, so we can cancel it if we're unassigned, and also listen to normal consumption (as already discussed, so we can ignore messages on unsubscribed partitions)

**Comment by dim** [comment]
@tcrayford please review my most recent [proposal above](https://github.com/bsm/sarama-cluster/issues/22#issuecomment-199692314). I think it would cover your use case too, but maybe I am missing something here.

**Comment by tcrayford** [comment]
@dim the "wait a configurable amount of time" is problematic here. Sometimes (when reloading state), my consumer will want to pause processing for a few minutes during a rebalance. Sometimes (when being told to work on a partition whose external state is already loaded), the wait time should be under 1ms. Having a single static configuration doesn't work at all. In my case I just listen to notifications, and mark a partition as "unassigned" when receiving work for it (and then just discard that message as it comes off the consumer channel)

I think the whole notion of "wait for the consumer to drain" is no good for this use case - you really want the person draining the channel to be able to say "yes I've drained it" to the goroutine doing the rebalancing. In my case, sometimes that'll be instant, sometimes it'll take a while.

None of this happens in the JVM 0.9 consumer right? There you'll just call `poll` and not get back messages from partitions you're unassigned from, which seems like it sidesteps all the issues here.

**Comment by dim** [comment]
@tcrayford apologies for the delay, "work" can be so distracting sometimes :) So I understand what you are trying to do, but I am not sure about the right API for it.

```go
select {
case msg := <- c.Messages():
   process(msg)
case rbl := <-c.Rebalance():    
  for msg := range rbl.Messages(): 
   process(msg)
  }
  rbl.Done()
}
```

The approach above would be prone to mistakes and I don't think it is very intuitive either. Thoughts?

**Comment by tcrayford** [comment]
@bsm that approach makes this kind of thing at least *possible* - currently you just have to hope the races aren't too bad or that sleeps don't get you down too much. I will give it some thought and get back to you.

**Comment by dim** [comment]
I had a few more thoughts and came up with two solutions, that I would like to put up for discussion, please see #38 and #39

**Comment by dim** [comment]
@tcrayford thoughts? cheers, dim

**Comment by tcrayford** [comment]
@dim sorry, been snowed under with work as of late. Let me look and I'll try and get back to you today.

**Comment by dim** [comment]
Closed via #39
